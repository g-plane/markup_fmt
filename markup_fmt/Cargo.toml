[package]
name = "markup_fmt"
version = "0.10.0"
edition = "2021"
authors = ["Pig Fang <g-plane@hotmail.com>"]
description = "Configurable HTML/Vue/Svelte/Astro/Angular/Jinja/Twig/Nunjucks/Vento formatter."
repository = "https://github.com/g-plane/markup_fmt"
license = "MIT"
exclude = ["/tests"]

[dependencies]
aho-corasick = "1.1"
css_dataset = "0.3"
itertools = "0.13"
memchr = "2.7"
once_cell = "1.19"
serde = { version = "1.0", optional = true }
tiny_pretty = { version = "0.2", features = ["unicode-width"] }

[dev-dependencies]
insta = { version = "1.39", features = ["glob"] }
similar-asserts = "1.5"
toml = "0.8"

[features]
config_serde = ["serde"]
