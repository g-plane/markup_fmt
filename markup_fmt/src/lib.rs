#![doc = include_str!("../README.md")]

mod ast;
pub mod config;
mod ctx;
mod error;
mod helpers;
mod parser;
mod printer;
mod state;

use crate::{config::FormatOptions, ctx::Ctx, parser::Parser, printer::DocGen, state::State};
pub use crate::{error::*, parser::Language};
use std::{borrow::Cow, path::Path};
use tiny_pretty::{IndentKind, PrintOptions};

/// Format the given source code.
///
/// An external formatter is required for formatting code
/// inside `<script>` or `<style>` tag.
/// If you don't need to format them or you don't have available formatters,
/// you can pass a closure that returns the original code. (see example below)
///
/// ```
/// use markup_fmt::{format_text, Language};
///
/// let code = r#"
/// <html>
///    <head>
///      <title>Example</title>
///      <style>button { outline: none; }</style>
///   </head>
///   <body><script>const a = 1;</script></body>
/// </html>"#;
///
/// let formatted = format_text(
///     code,
///     Language::Html,
///     &Default::default(),
///     |_, code, _| Ok::<_, std::convert::Infallible>(code.into()),
/// ).unwrap();
/// ```
///
/// For the external formatter closure,
///
/// - The first argument is file path.
/// Either script code or style code will be passed to this closure,
/// so you need to check file extension with the file path.
/// - The second argument is code that needs formatting.
/// - The third argument is print width that you may need to pass to formatter
/// if they support such option.
pub fn format_text<E, F>(
    code: &str,
    language: Language,
    options: &FormatOptions,
    external_formatter: F,
) -> Result<String, FormatError<E>>
where
    F: for<'a> FnMut(&Path, &'a str, usize) -> Result<Cow<'a, str>, E>,
{
    let mut parser = Parser::new(code, language.clone());
    let ast = parser.parse_root().map_err(FormatError::Syntax)?;
    let mut ctx = Ctx {
        language,
        indent_width: options.layout.indent_width,
        print_width: options.layout.print_width,
        options: &options.language,
        indent_level: 0,
        external_formatter,
        external_formatter_errors: Default::default(),
    };

    let doc = ast.doc(
        &mut ctx,
        &State {
            current_tag_name: None,
            is_root: true,
            in_svg: false,
        },
    );
    if !ctx.external_formatter_errors.is_empty() {
        return Err(FormatError::External(ctx.external_formatter_errors));
    }

    Ok(tiny_pretty::print(
        &doc,
        &PrintOptions {
            indent_kind: if options.layout.use_tabs {
                IndentKind::Tab
            } else {
                IndentKind::Space
            },
            line_break: options.layout.line_break.clone().into(),
            width: options.layout.print_width,
            tab_size: options.layout.indent_width,
        },
    ))
}

/// Detect language from file extension.
pub fn detect_language(path: impl AsRef<Path>) -> Option<Language> {
    let path = path.as_ref();
    match path.extension().and_then(std::ffi::OsStr::to_str) {
        Some("html") => {
            if path
                .file_stem()
                .map(|file_stem| file_stem.to_string_lossy().ends_with(".component"))
                .unwrap_or_default()
            {
                Some(Language::Angular)
            } else {
                Some(Language::Html)
            }
        }
        Some("vue") => Some(Language::Vue),
        Some("svelte") => Some(Language::Svelte),
        Some("astro") => Some(Language::Astro),
        Some("jinja" | "jinja2" | "twig" | "njk") => Some(Language::Jinja),
        Some("vto") => Some(Language::Vento),
        _ => None,
    }
}
