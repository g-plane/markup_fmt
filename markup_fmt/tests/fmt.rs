use insta::{Settings, assert_snapshot, glob};
use markup_fmt::{Language, config::FormatOptions, detect_language, format_text};
use std::{collections::HashMap, fs, path::Path};

#[test]
fn fmt_snapshot() {
    let pattern = "fmt/**/*.{html,vue,svelte,astro,jinja,njk,vto,mustache,xml}";
    glob!(pattern, |path| {
        let input = fs::read_to_string(path).unwrap();
        let language = detect_language(path).unwrap();

        let options = fs::read_to_string(path.with_file_name("config.toml"))
            .map(|config_file| {
                toml::from_str::<HashMap<String, FormatOptions>>(&config_file).unwrap()
            })
            .ok();

        if let Some(options) = options {
            options.into_iter().for_each(|(option_name, options)| {
                let output = run_format_test(path, &input, &options, language.clone());
                build_settings(path).bind(|| {
                    let name = path.file_stem().unwrap().to_str().unwrap();
                    assert_snapshot!(format!("{name}.{option_name}"), output);
                });
            })
        } else {
            let output = run_format_test(path, &input, &Default::default(), language);
            build_settings(path).bind(|| {
                let name = path.file_stem().unwrap().to_str().unwrap();
                assert_snapshot!(name, output);
            });
        }
    });
}

fn run_format_test(
    path: &Path,
    input: &str,
    options: &FormatOptions,
    language: Language,
) -> String {
    let output = format_text(&input, language.clone(), &options, |code, _| {
        Ok::<_, ()>(code.into())
    })
    .map_err(|err| format!("failed to format '{}': {:?}", path.display(), err))
    .unwrap();
    let regression_format = format_text(&output, language, &options, |code, _| {
        Ok::<_, ()>(code.into())
    })
    .map_err(|err| {
        format!(
            "syntax error in stability test '{}': {:?}",
            path.display(),
            err
        )
    })
    .unwrap();
    similar_asserts::assert_eq!(
        output,
        regression_format,
        "'{}' format is unstable",
        path.display()
    );

    output
}

fn build_settings(path: &Path) -> Settings {
    let mut settings = Settings::clone_current();
    settings.set_snapshot_path(path.parent().unwrap());
    settings.remove_snapshot_suffix();
    settings.set_prepend_module_to_snapshot(false);
    settings.remove_input_file();
    settings.set_omit_expression(true);
    settings.remove_input_file();
    settings.remove_info();
    settings
}
