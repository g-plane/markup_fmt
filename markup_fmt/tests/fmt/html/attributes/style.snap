---
source: markup_fmt/tests/fmt.rs
---
<div
  style="
  
    color:
    #fFf
  
  "
>
</div>

<div style=""></div>
<div style></div>

<div
  style="
  
    all: initial;display: block;
    contain: content;text-align: center;
  
  
  
    background: linear-gradient(to left, hotpink, #FFF00F, #ccc, hsla(240, 100%, 50%, .05), transparent);
    background: linear-gradient(to left, hsla(240, 100%, 50%, .05), red);
    max-width: 500px;margin: 0 auto;
    border-radius: 8px;transition: transform .2s ease-out;
  
  "
>
</div>

<div
  style="
    background: linear-gradient(to left, hotpink, hsla(240, 100%, 50%, .05), transparent);
  "
>
</div>

<div
  style="color : red;
         display    :inline "
>
</div>

<div
  style="
  
    color: green;
  
    display: inline 
  
  "
>
</div>

<div
  attribute-1
  attribute-2
  attribute-3
  attribute-4
  attribute-5
  attribute-6
  attribute-7
  style="css-prop-1: css-value;css-prop-2: css-value;css-prop-3: css-value;css-prop-4: css-value;"
  attribute-1
  attribute-2
  attribute-3
  attribute-4
  attribute-5
  attribute-6
  attribute-7
>
</div>

<div style="{{ ...styles }}"></div>

<div style="color: red; {{ otherStyles }}"></div>

<!-- Compact style argument -->
<div
  style="left: 50%;
         bottom: 0"
  class="absolute -translate-x-1/2 data-tag data-tag--lg data-tag--borders tint-green text-24 h-36"
>
  style is first arg
</div>

<div
  class="flex flex-col items-center absolute z-10 {% if title|length > 30 %}w-4/5{% else %}w-1/2{% endif %}"
  style="top:60%;
         transform:translate(0,-50%)"
>
  style is second arg
</div>

<div
  class="flex flex-col items-center absolute z-10 {% if title|length > 30 %}w-4/5{% else %}w-1/2{% endif %}"
  style="top:60%;
         transform:translate(0,-50%)"
  id="dfff"
>
  style is middle arg
</div>

<div
  style="top:60%;
         transform:translate(0,-50%)"
>
</div>

<div
  class="absolute right-0 bg-blue-020 rounded-full"
  style="height: 585px;
         width: 585px;
         filter: blur(86px);
         top: 153px"
>
  Big multiline style arg
</div>

<div
  style="top:90%;
         transform:translate(0,-50%)"
>
  <div
    style="top:80%;
           transform:translate(0,-50%)"
  >
    <div
      style="top:70%;
             transform:translate(0,-50%)"
    >
      Nested divs with multiline style attrs
    </div>
  </div>
</div>

<div
  style="color : red;
         display    :inline "
>
</div>
<div
  style="color : red;
         display    :inline; height: auto;
         position: absolute;
         top: 0;
         left: 0;
         right: 0;
         bottom: 0; "
>
</div>

<div
  style="top:60%;
         transform:translate(0,-50%)"
>
  <div
    style="top:60%;
           transform:translate(0,-50%)"
  >
    <div
      style="top:60%;
             transform:translate(0,-50%)"
    >
      Nested divs with odd off
    </div>
  </div>
</div>

<!-- Preserve multiline style argument -->
<div
  style="
    color: red;
    display: inline;
    height: auto;
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
  "
>
</div>
